### PR TITLE
Add BinSkim support to module

### PR DESCRIPTION
**Pull request by TravisEz13**

### CODE REVIEW

File: src/PSPackageProject.psd1 [hunk 1]
```diff
@@ -74,7 +74,8 @@ RequiredModules = @(
 
 # Functions to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no functions to export.
 FunctionsToExport = @(
-    'New-PSPackageProjectHelp', 'Invoke-PSPackageProjectTest'
+    'New-PSPackageProjectHelp'
```
Comment (rjmholt) [inline src/PSPackageProject.psd1]: Do the new functions need to be exported here?

File: src/PSPackageProject.psm1 [hunk 2]
```diff
@@ -11,4 +11,100 @@ function Invoke-PSPackageProjectTest
...
+    $packageName = 'microsoft.codeanalysis.binskim'
+    $packageLocation = Join-Path -Path ([System.io.path]::GetTempPath()) -ChildPath 'pspackageproject-packages'
+    Write-Verbose "Finding binskim..." -Verbose
+    $packageInfo = Find-Package -Name $packageName -Source $sourceName
+    if($IsLinux)
+    {
+        $binaryName ='BinSkim'
+        $rid = 'linux-x64'
+    }
+    elseif($IsWindows)
```
Comment (rjmholt) [inline src/PSPackageProject.psm1]: Should this work in PS 5.1 (`$IsWindows -ne $false`)?

File: src/PSPackageProject.psm1 [hunk 3]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
...
+        [parameter(Mandatory)]
+        [string]
+        $Title,
+        [string]
+        $Type = 'NUnit'
+    )
+
+    $artifactPath = (Resolve-Path $Path).ProviderPath
+
+    if($env:TF_BUILD)
```
Comment (rjmholt) [inline src/PSPackageProject.psm1]: If this function only is supposed to work in Azure DevOps, should this be:

```
if (-not $env:TF_BUILD)
{
    Write-Warning "something about only working in Azure Devops"
}

# Content
```

File: src/PSPackageProject.psm1 [hunk 4]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
...
+    else
+    {
+        $toAnalyze = Join-Path -Path $packageLocation -ChildPath $dirName -AdditionalChildPath 'tools', 'netcoreapp2.0', $rid, '*'
+    }
+
+    $outputPath =  Join-Path -Path ([System.io.path]::GetTempPath()) -ChildPath 'pspackageproject-results.json'
+    Write-Verbose "Running binskim..." -Verbose
+    & $toolLocation analyze $toAnalyze --output $outputPath --pretty-print  > binskim.log 2>&1
+
+    $testsPath =Join-Path -Path $PSScriptRoot -ChildPath 'tasks' -AdditionalChildPath 'BinSkim', 'binskim.tests.ps1'
```
Comment (rjmholt) [inline src/PSPackageProject.psm1]:
```suggestion
    $testsPath = Join-Path -Path $PSScriptRoot -ChildPath 'tasks' -AdditionalChildPath 'BinSkim', 'binskim.tests.ps1'
```

**Review by rjmholt — APPROVED** [review]

File: src/PSPackageProject.psm1 [hunk 5]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
     )
 
     ## TODO implement calling tests
+    Invoke-BinSkim
+}
+
+function Invoke-BinSkim
```
Comment (rjmholt) [inline src/PSPackageProject.psm1]: Do these functions need to be exported in the manifest?

File: src/PSPackageProject.psm1 [hunk 6]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
     )
 
     ## TODO implement calling tests
+    Invoke-BinSkim
```
Comment (adityapatwardhan) [inline src/PSPackageProject.psm1]: This should be called when `$Type -eq 'StaticAnalysis'`

File: src/PSPackageProject.psm1 (hunk 5 above)
Comment (adityapatwardhan) [reply src/PSPackageProject.psm1]: I think, no. `-Type StaticAnalysis` should be the "public" interface.

File: src/PSPackageProject.psm1 [hunk 7]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
...
+    elseif($IsWindows)
+    {
+        $binaryName ='BinSkim.exe'
+        if([Environment]::Is64BitOperatingSystem)
+        {
+            $rid = 'win-x64'
+        }
+        else
+        {
+            $rid = 'win-x86'
```
Comment (adityapatwardhan) [inline src/PSPackageProject.psm1]: Do we need this? AzDO does not have a x86 agent.

File: src/PSPackageProject.psd1 (hunk 1 above)
Comment (TravisEz13) [reply src/PSPackageProject.psd1]: no

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/PSPackageProject.psm1 (hunk 3 above)
Comment (TravisEz13) [reply src/PSPackageProject.psm1]: We just don’t do this if we aren’t on azure dev Ops.  no need for a warning.

File: src/PSPackageProject.psm1 (hunk 5 above)
Comment (TravisEz13) [reply src/PSPackageProject.psm1]: No

File: src/PSPackageProject.psm1 (hunk 7 above)
Comment (TravisEz13) [reply src/PSPackageProject.psm1]: We should allow this locally

File: src/PSPackageProject.psm1 (hunk 2 above)
Comment (TravisEz13) [reply src/PSPackageProject.psm1]:
```suggestion
    elseif($IsWindows -ne $false)
```

File: src/PSPackageProject.psm1 (hunk 3 above)
Comment (TravisEz13) [inline src/PSPackageProject.psm1]:
```suggestion
    # Just do nothing if we are not in AzDevOps
    if($env:TF_BUILD)
```

File: src/PSPackageProject.psm1 (hunk 6 above)
Comment (TravisEz13) [reply src/PSPackageProject.psm1]:
```suggestion
    if ($Type -eq 'StaticAnalysis')
    {
        Invoke-BinSkim
    }
```

File: src/PSPackageProject.psm1 [hunk 8]
```diff
@@ -93,6 +93,102 @@ function Invoke-PSPackageProjectTest
...
+        $Title,
+        [string]
+        $Type = 'NUnit'
+    )
+
+    $artifactPath = (Resolve-Path $Path).ProviderPath
+
+    if($env:TF_BUILD)
+    {
+        Write-Host "##vso[results.publish type=$Type;mergeResults=true;runTitle=$Title;publishRunAttachments=true;resultFiles=$artifactPath;]"
```
Comment (TravisEz13) [inline src/PSPackageProject.psm1]:
```suggestion
        Write-Host "##vso[results.publish type=$Type;mergeResults=true;runTitle=$Title;publishRunAttachments=true;resultFiles=$artifactPath;failTaskOnFailedTests=true;]"
```

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**